File: 3. Behavioral Design Patterns/State/Go/StateTemplate.go
package main

// State Pattern – Go Template

type State interface{ Handle(*Context) }

type Context struct{ state State }
func NewContext(s State) *Context { return &Context{state: s} }
func (c *Context) SetState(s State) { c.state = s }
func (c *Context) Request() { c.state.Handle(c) }

type ConcreteStateA struct{}
func (s *ConcreteStateA) Handle(c *Context) { /* work */ c.SetState(&ConcreteStateB{}) }

type ConcreteStateB struct{}
func (s *ConcreteStateB) Handle(c *Context) { /* work */ c.SetState(&ConcreteStateA{}) }

func main(){ ctx := NewContext(&ConcreteStateA{}); ctx.Request(); ctx.Request() }

File: 3. Behavioral Design Patterns/State/Python/StateTemplate.py
# State Pattern – Python Template

from abc import ABC, abstractmethod

class State(ABC):
    @abstractmethod
    def handle(self, context):
        pass

class Context:
    def __init__(self, initial_state):
        self.state = initial_state
    
    def set_state(self, state):
        self.state = state
    
    def request(self):
        self.state.handle(self)

class ConcreteStateA(State):
    def handle(self, context):
        print("ConcreteStateA handling request")
        context.set_state(ConcreteStateB())

class ConcreteStateB(State):
    def handle(self, context):
        print("ConcreteStateB handling request")
        context.set_state(ConcreteStateA())

def main():
    context = Context(ConcreteStateA())
    context.request()
    context.request()

if __name__ == "__main__":
    main()
